File: 프로그래머스/프로그래머스_Lv3/디스크 컨트롤러/디스크 컨트롤러.java
import java.util.Arrays;
import java.util.PriorityQueue;

class Solution {


    private int answer = 0;
    private int time = 0;
    private PriorityQueue<int[]> workQue = new PriorityQueue<>((o1, o2) -> o1[1] - o2[1]);
    private PriorityQueue<int[]> inQue = new PriorityQueue<>((o1, o2) -> o1[0] - o2[0]);

    public int solution(int[][] jobs) {
        for(int[] j : jobs) inQue.offer(j);
        while(hasWorksToDo()){
            while(canWorkNow()) workQue.offer(inQue.poll());
            setTime();
        }
        answer /= jobs.length;
        return answer;
    }
    
    private boolean canWorkNow() {
        return !inQue.isEmpty() && inQue.peek()[0] <= time;
    }

    private boolean hasWorksToDo() {
        return !inQue.isEmpty() || !workQue.isEmpty();
    }

    private void setTime() {
        if(workQue.isEmpty()){ 
            time = inQue.peek()[0];
            return;
        }
        
        int[] j = workQue.poll();
        answer += time + j[1] - j[0];
        time += j[1];
    }
}

File: 프로그래머스/프로그래머스_Lv3/디스크 컨트롤러/디스크 컨트롤러.js
const solution = (jobs) => {
  const worksByWeight = [...jobs].sort((a, b) => a[1] - b[1]);
  const worksByStart = [...jobs].sort((a, b) => a[0] - b[0]);
  let time = 0;
  let answer = 0;
  while (worksByStart.length !== 0 && worksByWeight.length !== 0) {
    while (worksByWeight.length !== 0 && worksByStart[0][0] <= time) {
      if (worksByWeight.length === 0) {
        time = worksByStart[0][0];
        continue;
      }

      const j = worksByWeight.shift();
      answer += time + j[1] - j[0];
      time += j[1];
    }
  }
  return answer / jobs.length;
};
